File: runtime/src/iree/base/internal/wait_handle_emscripten.c
// Copyright 2023 The IREE Authors
//
// Licensed under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

// clang-format off: must be included before all other headers.
#include "iree/base/internal/wait_handle_impl.h"
// clang-format on

#include "iree/base/api.h"
#include "iree/base/internal/wait_handle.h"

// This implementation for the web platform via Emscripten uses JavaScript
// Promise objects for asynchronous waiting:
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise
//
// Synchronous wait APIs (e.g. iree_wait_one) are not supported, as using
// Promises requires interacting with the asynchronous browser event loop.
//   * Note: JSPI (https://v8.dev/blog/jspi) and Asyncify
//     (https://emscripten.org/docs/porting/asyncify.html)
//     could probably be used to interop with this, but the real goal of this
//     implementation is browser-native asynchronous behavior.
//
// Wait handles may be asynchronously waited on via the JavaScript API:
//   ```
//   // C
//   int handle = iree_wait_primitive_promise_create(false);
//
//   // JS
//   const promiseWrapper = IreeWaitHandlePromise.getPromiseWrapper(handle);
//   promiseWrapper.promise.then(() => { ... });
//
//   // C
//   iree_wait_primitive_promise_set(handle);
//   ```
#if IREE_WAIT_API == IREE_WAIT_API_PROMISE && defined(IREE_PLATFORM_EMSCRIPTEN)

#include <emscripten.h>

//===----------------------------------------------------------------------===//
// externs from wait_handle_emscripten.js
//===----------------------------------------------------------------------===//

extern int iree_wait_primitive_promise_create(bool initial_state);
extern void iree_wait_primitive_promise_delete(int promise_handle);
extern void iree_wait_primitive_promise_set(int promise_handle);
extern void iree_wait_primitive_promise_reset(int promise_handle);

//===----------------------------------------------------------------------===//
// iree_wait_primitive_* raw calls
//===----------------------------------------------------------------------===//

void iree_wait_handle_close(iree_wait_handle_t* handle) {
  switch (handle->type) {
#if defined(IREE_HAVE_WAIT_TYPE_JAVASCRIPT_PROMISE)
    case IREE_WAIT_PRIMITIVE_TYPE_JAVASCRIPT_PROMISE: {
      iree_wait_primitive_promise_delete(handle->value.promise.handle);
      break;
    }
#endif  // IREE_HAVE_WAIT_TYPE_JAVASCRIPT_PROMISE
    default:
      break;
  }
  iree_wait_handle_deinitialize(handle);
}

iree_status_t iree_wait_one(iree_wait_handle_t* handle,
                            iree_time_t deadline_ns) {
  return iree_make_status(IREE_STATUS_UNIMPLEMENTED,
                          "wait_one unimplemented on Emscripten");
}

//===----------------------------------------------------------------------===//
// iree_event_t
//===----------------------------------------------------------------------===//

iree_status_t iree_event_initialize(bool initial_state,
                                    iree_event_t* out_event) {
  memset(out_event, 0, sizeof(*out_event));
  out_event->type = IREE_WAIT_PRIMITIVE_TYPE_JAVASCRIPT_PROMISE;

  int promise_handle = iree_wait_primitive_promise_create(initial_state);
  out_event->value.promise.handle = promise_handle;
  return iree_ok_status();
}

void iree_event_deinitialize(iree_event_t* event) {
  iree_wait_handle_close(event);
}

void iree_event_set(iree_event_t* event) {
  if (!event) return;
  iree_wait_primitive_promise_set(event->value.promise.handle);
}

void iree_event_reset(iree_event_t* event) {
  if (!event) return;
  iree_wait_primitive_promise_reset(event->value.promise.handle);
}

#endif  // IREE_WAIT_API == IREE_WAIT_API_INPROC &&
        // defined(IREE_PLATFORM_EMSCRIPTEN)

File: runtime/src/iree/base/internal/wait_handle_emscripten.js
// Copyright 2023 The IREE Authors
//
// Licensed under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

// This is the JavaScript side of wait_handle_emscripten.c
//
// Each `iree_wait_handle_t` tracks a "iree_wait_primitive".
// This Emscripten implementation is backed by Promises and it
//   * creates a "promise wrapper" for each active wait handle, where each
//     wrapper includes a Promise object and supporting data
//   * returns an opaque handle used to reference that wrapper

const LibraryIreeWaitHandleEmscripten = {
  $IreeWaitHandleEmscripten: {
    // Counter for opaque handles, shared across the entire module/process/page.
    // Note: start at 1, leaving 0 as a sentinel for uninitialized.
    _nextPromiseHandle: 1,

    // Dictionary of promiseHandles -> promise wrapper objects:
    // {
    //   promise:    Promise object
    //   isSettled:  Bool
    //   resolve:    Function that resolves |promise|
    //   reject:     Function that rejects  |promise|
    // }
    // TODO(scotttodd): use a class for that ^
    _promiseWrappers: {},

    _createPromiseWrapper: function(promiseHandle, initialState) {
      if (initialState) {
        // Start resolved (settled).
        const resolvedPromise = Promise.resolve();
        const promiseWrapper = {
          promise: resolvedPromise,
          isSettled: true,
        };
        IreeWaitHandleEmscripten._promiseWrappers[promiseHandle] =
            promiseWrapper;
      } else {
        // Start pending, track the resolve and reject functions.
        const pendingPromise = new Promise((resolve, reject) => {
          const promiseWrapper = {
            promise: undefined,
            isSettled: false,
            resolve: resolve,
            reject: reject,
          };
          IreeWaitHandleEmscripten._promiseWrappers[promiseHandle] =
              promiseWrapper;
        });
        IreeWaitHandleEmscripten._promiseWrappers[promiseHandle].promise =
            pendingPromise;
      }
    },

    getPromise: function(promiseHandle) {
      return IreeWaitHandleEmscripten._promiseWrappers[promiseHandle].promise;
    },
  },

  iree_wait_primitive_promise_create: function(initialState) {
    const promiseHandle = IreeWaitHandleEmscripten._nextPromiseHandle++;
    IreeWaitHandleEmscripten._createPromiseWrapper(promiseHandle, initialState);
    return promiseHandle;
  },

  iree_wait_primitive_promise_delete: function(promiseHandle) {
    const promiseWrapper =
        IreeWaitHandleEmscripten._promiseWrappers[promiseHandle];
    if (!promiseWrapper.isSettled && promiseWrapper.reject !== undefined) {
      promiseWrapper.reject();
      promiseWrapper.isSettled = true;
    }
    delete IreeWaitHandleEmscripten._promiseWrappers[promiseHandle];
  },

  iree_wait_primitive_promise_set: function(promiseHandle) {
    const promiseWrapper =
        IreeWaitHandleEmscripten._promiseWrappers[promiseHandle];
    if (promiseWrapper.resolve !== undefined) {
      promiseWrapper.resolve();
      promiseWrapper.isSettled = true;
    }
  },

  iree_wait_primitive_promise_reset: function(promiseHandle) {
    const promiseWrapper =
        IreeWaitHandleEmscripten._promiseWrappers[promiseHandle];

    // No-op if already unsignaled.
    if (!promiseWrapper.isSettled) return;

    // Promises are are permanently resolved and can't be 'reset', so create a
    // new wrapper using the same handle.
    // Since the previous Promise was resolved or rejected already, listeners
    // should have already been notified and we aren't leaving any orphaned.
    // (?) This synchronizes on the browser event loop, so there are no races.
    IreeWaitHandleEmscripten._createPromiseWrapper(promiseHandle, false);
  },
}

autoAddDeps(LibraryIreeWaitHandleEmscripten, '$IreeWaitHandleEmscripten');
mergeInto(LibraryManager.library, LibraryIreeWaitHandleEmscripten);
